metricbeat/module/system/network/network.go

```diff
@@ -104,46 +105,102 @@ func (m *MetricSet) Fetch(r mb.ReporterV2) error {
 			MetricSetFields: ioCountersToMapStr(counters),
 		})
 
-		// accumulate values from all interfaces
-		networkInBytes += counters.BytesRecv
-		networkOutBytes += counters.BytesSent
-		networkInPackets += counters.PacketsRecv
-		networkOutPackets += counters.PacketsSent
+		// sum the values at a per-interface level
+		// Makes us less likely to overload a value somewhere.
+		prevCounters, ok := m.prevInterfaceCounter[counters.Name]
+		if !ok {
+			m.prevInterfaceCounter[counters.Name] = networkCounter{
+				NetworkInBytes:    counters.BytesRecv,
+				NetworkInPackets:  counters.PacketsRecv,
+				NetworkOutBytes:   counters.BytesSent,
+				NetworkOutPackets: counters.PacketsSent,
+			}
+			continue
+		}
+		// create current set of gauges
+		currentDiff := networkCounter{
+			NetworkInBytes:    createGaugeWithRollover(counters.BytesRecv, prevCounters.NetworkInBytes),
+			NetworkInPackets:  createGaugeWithRollover(counters.PacketsRecv, prevCounters.NetworkInPackets),
+			NetworkOutBytes:   createGaugeWithRollover(counters.BytesSent, prevCounters.NetworkOutBytes),
+			NetworkOutPackets: createGaugeWithRollover(counters.PacketsSent, prevCounters.NetworkOutPackets),
+		}
+
+		m.currentGaugeCounter[counters.Name] = currentDiff
+
+		m.prevInterfaceCounter[counters.Name] = networkCounter{
+			NetworkInBytes:    counters.BytesRecv,
+			NetworkInPackets:  counters.PacketsRecv,
+			NetworkOutBytes:   counters.BytesSent,
+			NetworkOutPackets: counters.PacketsSent,
+		}
 
 		if !isOpen {
 			return nil
 		}
 	}
 
-	if m.prevCounters != (networkCounter{}) {
-		// convert network metrics from counters to gauges
+	if len(m.currentGaugeCounter) != 0 {
+
+		var totalNetworkInBytes, totalNetworkInPackets, totalNetworkOutBytes, totalNetworkOutPackets uint64
+		for _, iface := range m.currentGaugeCounter {
+			totalNetworkInBytes += iface.NetworkInBytes
```

Comment:
Do we need to be concerned about rollover here? If we have several busy interfaces, I think we would rollover here before the individual interface rolls over.

Reply:
So, my thinking here is that rollover here would be fairly unlikely, since at this point we're summing the difference between gauges across time periods, also, I'm not sure how we would "fix" a rollover at that point, since it would mean we need to transmit data in a >64bit type.

Reply:
I agree it is unlikely.  We could just log if the total value is ever less than the what we were += into it.  A log message makes debugging much easier if we ever do hit this edge case.